### PR TITLE
Refactor LightningDataParallel

### LINKED ISSUE #5506: Keeping DP override in sync with upstream torch
Same as in #4630, we want to redirect forward to the training_step, validaion_step, test_step methods but without having to copy the DP implementation directly from pytorch codebase. This would also allow users to easily override DP in Lightning.

See refactor for DDP here: #5185

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #5506 
Follow-up to #5185

It is a bit hard to see what happens in the diff, because there are a lot of deletions. 
_I recommend to view diff in split view: Go to Files changed tab -> settings wheel -> change unifided to split_

The refactor for DP here is basically analogeous to the refactor I did in #5185, with the addition that in DP we need to 1) convert Python scalars to tensors (to match functionality we had before) and 2) automatically unsqueeze 0-dim tensors 

Todo: write tests

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/legacy/dp_accelerator.py [hunk 1]
```diff
@@ -178,8 +178,10 @@ def reinit_scheduler_properties(self, optimizers: list, schedulers: list):
                     scheduler.load_state_dict(state)
 
     def get_reference_model(self, model) -> LightningModule:
-        if isinstance(model, LightningDataParallel):
-            return model.module
+        if isinstance(model, torch.nn.DataParallel):
+            model = model.module
+        if isinstance(model, LightningParallelModule):
+            model = model.module
```
Comment (ydcjeff) [inline pytorch_lightning/accelerators/legacy/dp_accelerator.py:184]: any difference between these two?

Comment (awaelchli) [reply pytorch_lightning/accelerators/legacy/dp_accelerator.py:184]: yep! there is a difference. The model is wrapped like this:

DataParallel(LightingParallelModule(BoringModel))

the outermost is just the torch DP
the middle one is ours, it takes care of forwarding "forward" to the correct step method
and the one in the inner most layer is just the LightningModule from the user.

Comment (awaelchli) [reply pytorch_lightning/accelerators/legacy/dp_accelerator.py:184]: therefore, to get to the LightningModule we potentially need to call `model.module.module`

File: pytorch_lightning/overrides/data_parallel.py [hunk 2]
```diff
@@ -45,133 +40,103 @@ def _find_tensors(obj):  # pragma: no-cover
...
+            output = self.module.test_step(*inputs, **kwargs)
+            warn_if_output_is_none(output, "test_step")
+        elif running_stage == RunningStage.EVALUATING:
+            output = self.module.validation_step(*inputs, **kwargs)
+            warn_if_output_is_none(output, "validation_step")
+        else:
+            output = self.module.predict(*inputs, **kwargs)
 
-            return outputs
+        return output
```
Comment (ananthsub) [inline pytorch_lightning/overrides/data_parallel.py:97]: n00b question: does data parallel work outside of training/testing?

File: pytorch_lightning/overrides/data_parallel.py [hunk 3]
```diff
@@ -45,133 +40,103 @@ def _find_tensors(obj):  # pragma: no-cover
...
+            DeprecationWarning
+        )
+        super().__init__(LightningDistributedModule(module), *args, **kwargs)
 
-        result = original_class()
 
-        result.update(outputs)
-        result['meta'] = meta
-        return result
+class LightningModuleWrapperBase(torch.nn.Module):
```
Comment (ananthsub) [inline pytorch_lightning/overrides/data_parallel.py:68]: Is this module intended to be private, used only for the data parallel overrides? If so, should we denote that by naming it `_LightningModuleWrapperBase`?

**Review by ananthsub — COMMENTED** [review]
excellent to see how much simpler the code is!

File: pytorch_lightning/overrides/data_parallel.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:97]: yes, outside of training and testing it behaves like the regular forward, i.e., it will pass the arguments to LightningModule.predict which defaults to redirecting to LightningModule.forward, unless user overrides that.

File: pytorch_lightning/overrides/data_parallel.py (hunk 3 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py:68]: yes, I think that makes sense

File: pytorch_lightning/overrides/data_parallel.py [hunk 4]
```diff
@@ -45,178 +40,132 @@ def _find_tensors(obj):  # pragma: no-cover
...
+            dtype=numbers.Number,
+            function=python_scalar_to_tensor,
+            device=self.module.device
+        )
+        output = apply_to_collection(
+            output,
+            dtype=torch.Tensor,
+            function=unsqueeze_scalar_tensor,
+        )
+        return output
```
Comment (justusschock) [inline pytorch_lightning/overrides/data_parallel.py:139]: can we maybe combine these functions? then we only need to traverse the collections once

File: pytorch_lightning/overrides/data_parallel.py [hunk 5]
```diff
@@ -45,178 +40,132 @@ def _find_tensors(obj):  # pragma: no-cover
...
+            )
+
+        Args:
+            pl_module: the model to wrap
+
+        """
+        super().__init__(pl_module)
+
+    def forward(self, *inputs, **kwargs):
+        return super().forward(*inputs, **kwargs)
```
Comment (justusschock) [inline pytorch_lightning/overrides/data_parallel.py:166]:
```suggestion
```
the super method will be used automatically if not overwritten...

File: pytorch_lightning/overrides/data_parallel.py [hunk 6]
```diff
@@ -45,178 +40,132 @@ def _find_tensors(obj):  # pragma: no-cover
...
+        )
+        output = apply_to_collection(
+            output,
+            dtype=torch.Tensor,
+            function=unsqueeze_scalar_tensor,
+        )
+        return output
+
+
+class LightningDistributedModule(_LightningModuleWrapperBase):
```
Comment (justusschock) [inline pytorch_lightning/overrides/data_parallel.py:142]: can we move this to a different file, coming from plain PyTorch, I wouldn't look into data_parallel.py if looking for distributed stuff

File: pytorch_lightning/overrides/data_parallel.py [hunk 7]
```diff
@@ -45,178 +40,132 @@ def _find_tensors(obj):  # pragma: no-cover
...
-            if isinstance(result, torch.Tensor):
-                return result
-    if isinstance(obj, dict):
-        for result in map(get_a_var, obj.items()):
-            if isinstance(result, torch.Tensor):
-                return result
-    return None
-
-
 warning_cache = WarningCache()
```
Comment (justusschock) [inline pytorch_lightning/overrides/data_parallel.py:43]: do we really want a global variable for this. what about creating one for each running stage inside the module?

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/overrides/data_parallel.py (hunk 5 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py]: yes, I should know that xD

File: pytorch_lightning/overrides/data_parallel.py (hunk 6 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py]: I moved it and also the base class to a separate file. Let's see, is it better now?

File: pytorch_lightning/overrides/data_parallel.py (hunk 7 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/data_parallel.py]: yes, that's reasonable. I will follow up with a PR on this and keeping it a refactor here.

**Review by ananthsub — APPROVED** [review]
LGTM!

File: pytorch_lightning/overrides/base.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+import torch
+
+from pytorch_lightning import LightningModule
+from pytorch_lightning.trainer.states import RunningStage
+from pytorch_lightning.utilities.warnings import WarningCache
+
+warning_cache = WarningCache()
+
+
+class _LightningModuleWrapperBase(torch.nn.Module):
```
Comment (Borda) [inline pytorch_lightning/overrides/base.py:25]: interesting I think I have never seen "protected" class :]

File: pytorch_lightning/overrides/data_parallel.py [hunk 9]
```diff
@@ -11,363 +11,90 @@
...
-
-    def parallel_apply(self, replicas, inputs, kwargs):
-        return parallel_apply(replicas, inputs, kwargs, self.device_ids[:len(replicas)])
+    def __init__(self, module: LightningModule, *args, **kwargs):
+        warnings.warn(
+            "The usage of `LightningDataParallel` is deprecated since v1.2 and will be removed in v1.4."
+            " From now on we recommend to directly subclass `torch.nn.parallel.DataParallel`.",
+            DeprecationWarning
+        )
+        super().__init__(LightningParallelModule(module), *args, **kwargs)
```
Comment (Borda) [inline pytorch_lightning/overrides/data_parallel.py:36]: :+1:

File: tests/deprecated_api/test_remove_1-4.py [hunk 10]
```diff
@@ -180,3 +187,14 @@ def test_v1_4_0_deprecated_lightning_distributed_data_parallel(tmpdir):
...
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="test requires GPU machine")
+def test_v1_4_0_deprecated_lightning_data_parallel():
+    model = BoringModel()
+    with pytest.deprecated_call(
+            match="`LightningDataParallel` is deprecated since v1.2 and will be removed in v1.4."
+    ):
+        dp_model = LightningDataParallel(model, device_ids=[0])
+        assert isinstance(dp_model, torch.nn.DataParallel)
+        assert isinstance(dp_model.module, LightningParallelModule)
```
Comment (Borda) [inline tests/deprecated_api/test_remove_1-4.py:200]:
```suggestion
    assert isinstance(dp_model, torch.nn.DataParallel)
    assert isinstance(dp_model.module, LightningParallelModule)
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/overrides/base.py (hunk 8 above)
Comment (awaelchli) [reply pytorch_lightning/overrides/base.py:25]: An example that acomes to mind is _LRScheduler from PyTorch :)

**Comment by awaelchli** [comment]
Thanks for the reviews. 
@Borda Unfortunately, there is an existing DP test failing which I haven't been able to figure out yet, so the PR is not yet ready to merge. So I am removing the ready label for now

**Comment by awaelchli** [comment]
@Borda Tests passing now, and I fixed the cyclic import error. This should be good now :)
